src/sysdolphin/baselib/aobj.c

```diff
@@ -309,112 +310,79 @@ void RObjForeachAnim(HSD_RObj* robj, s32 flags, void* r5, s32 r6, void* r7)
-lbl_8036487C:
-    /* 8036487C 0036145C  2C 18 00 00 */	cmpwi r24, 0
-    /* 80364880 00361460  41 82 00 28 */	beq lbl_803648A8
-    /* 80364884 00361464  80 7B 00 64 */	lwz r3, 0x64(r27)
-    /* 80364888 00361468  28 03 00 00 */	cmplwi r3, 0
-    /* 8036488C 0036146C  41 82 00 1C */	beq lbl_803648A8
-    /* 80364890 00361470  38 9B 00 00 */	addi r4, r27, 0
-    /* 80364894 00361474  38 DD 00 00 */	addi r6, r29, 0
-    /* 80364898 00361478  38 FE 00 00 */	addi r7, r30, 0
-    /* 8036489C 0036147C  39 1F 00 00 */	addi r8, r31, 0
-    /* 803648A0 00361480  38 A0 00 0B */	li r5, 0xb
-    /* 803648A4 00361484  4B FF FD 35 */	bl callbackForeachFunc
-lbl_803648A8:
-    /* 803648A8 00361488  83 7B 00 08 */	lwz r27, 8(r27)
-lbl_803648AC:
-    /* 803648AC 0036148C  28 1B 00 00 */	cmplwi r27, 0
-    /* 803648B0 00361490  40 82 FF CC */	bne lbl_8036487C
-lbl_803648B4:
-    /* 803648B4 00361494  2C 17 00 00 */	cmpwi r23, 0
-    /* 803648B8 00361498  80 9C 00 0C */	lwz r4, 0xc(r28)
-    /* 803648BC 0036149C  41 82 00 48 */	beq lbl_80364904
-    /* 803648C0 003614A0  28 04 00 00 */	cmplwi r4, 0
-    /* 803648C4 003614A4  41 82 00 40 */	beq lbl_80364904
-    /* 803648C8 003614A8  A0 04 00 0C */	lhz r0, 0xc(r4)
-    /* 803648CC 003614AC  54 00 04 A6 */	rlwinm r0, r0, 0, 0x12, 0x13
-    /* 803648D0 003614B0  2C 00 10 00 */	cmpwi r0, 0x1000
-    /* 803648D4 003614B4  40 82 00 30 */	bne lbl_80364904
-    /* 803648D8 003614B8  80 64 00 14 */	lwz r3, 0x14(r4)
-    /* 803648DC 003614BC  28 03 00 00 */	cmplwi r3, 0
-    /* 803648E0 003614C0  41 82 00 24 */	beq lbl_80364904
-    /* 803648E4 003614C4  80 63 00 20 */	lwz r3, 0x20(r3)
-    /* 803648E8 003614C8  28 03 00 00 */	cmplwi r3, 0
-    /* 803648EC 003614CC  41 82 00 18 */	beq lbl_80364904
-    /* 803648F0 003614D0  38 DD 00 00 */	addi r6, r29, 0
-    /* 803648F4 003614D4  38 FE 00 00 */	addi r7, r30, 0
-    /* 803648F8 003614D8  39 1F 00 00 */	addi r8, r31, 0
-    /* 803648FC 003614DC  38 A0 00 09 */	li r5, 9
-    /* 80364900 003614E0  4B FF FC D9 */	bl callbackForeachFunc
-lbl_80364904:
-    /* 80364904 003614E4  83 9C 00 04 */	lwz r28, 4(r28)
-lbl_80364908:
-    /* 80364908 003614E8  28 1C 00 00 */	cmplwi r28, 0
-    /* 8036490C 003614EC  40 82 FF 04 */	bne lbl_80364810
-    /* 80364910 003614F0  BA E1 00 24 */	lmw r23, 0x24(r1)
-    /* 80364914 003614F4  80 01 00 4C */	lwz r0, 0x4c(r1)
-    /* 80364918 003614F8  38 21 00 48 */	addi r1, r1, 0x48
-    /* 8036491C 003614FC  7C 08 03 A6 */	mtlr r0
-    /* 80364920 00361500  4E 80 00 20 */	blr 
-} // clang-format on
-#pragma pop
+void PObjForeachAnim(HSD_PObj* pobj, s32 flags, void* r5, s32 r6, void* r7)
+{
+    if (pobj == NULL)
+        return;
+
+    if ((pobj->flags & 0x3000) == 0x1000
+        && pobj->u.x14_unk != NULL && pobj->u.x14_unk->aobj != NULL) {
+        callbackForeachFunc(pobj->u.x14_unk->aobj, pobj, POBJ_TYPE, r5, r6, r7);
+    }
+}
+
+void MObjForeachAnim(HSD_MObj* mobj, s32 flags, void* r5, s32 r6, void* r7)
```

Comment:
`void PObjForeachAnim(HSD_PObj *obj, HSD_TypeMask mask, void (*func)(void), AObj_Arg_Type arg_type, callbackArg *arg);`

Reply:
`typedef enum _HSD_TypeMask
{
	AOBJ_BIT      = 1 << 0,
	COBJ_BIT      = 1 << 1,
	DOBJ_BIT      = 1 << 2,
	FOBJ_BIT      = 1 << 3,
	FOG_BIT       = 1 << 4,
	JOBJ_BIT      = 1 << 5,
	LOBJ_BIT      = 1 << 6,
	MOBJ_BIT      = 1 << 7,
	POBJ_BIT      = 1 << 8,
	ROBJ_BIT      = 1 << 9,
	TOBJ_BIT      = 1 << 10,
	WOBJ_BIT      = 1 << 11,
	RENDER_BIT    = 1 << 12,
	CHAN_BIT      = 1 << 13,
	TEVREG_BIT    = 1 << 14,
	CBOBJ_BIT     = 1 << 15,
	ALL_TYPE_BITS = 1 << 16
} HSD_TypeMask;
`

Reply:
Should that enum def go in the aobj.h header?

Reply:
Oh, this is very similar to HSD_Type, so I'll put it next to that enum in object.h.